src/realm/object-store/shared_realm.hpp

```diff
@@ -378,6 +378,22 @@ class Realm : public std::enable_shared_from_this<Realm> {
         return !m_transaction && !m_coordinator;
     }
 
+    /**
+     * Deletes the following files for this Realm's configuration if they exist:
+     * - the Realm file itself
+     * - the .log file and its legacy versions: .log_a and .log_b
+     * - the .note file
+     * - the .management folder
+     *
+     * The .lock file for this Realm cannot and will not be deleted as this is unsafe.
+     * If a different process / thread is accessing the realm at the same time a corrupt state
+     * could be the result and checking for a single process state is not possible here.
+     *
+     * @return Yes, if all files could either be deleted or did not exist.
+     *         No if at least one existing file coule not be deleted.
```

Comment:
The comment on the return value is not accurate. Currently you are using `File::try_remove()` and ignoring its return value. Also consider if you'd like to have this function throw on permission denied errors or if you'd prefer to catch exceptions and return false if anything throws. Or if the behaviour is as intended, you could update this comment to state that it may throw on error, and that there is no guarantee that the files will be deleted.

Reply:
1. The difference between `File::try_remove()` and `File::remove()` is that the latter throws an error if the file did not exist whereas the former just returns `false`. The current SDKs implementations ignore non-existing files in general.
Using `File::try_remove()` and ignoring the `false` case (while still throwing other exception) seemed to me the closed mirror of the SDK implementations.
So if I understand you correctly, the only thing missing here is to mention that we also return `false` is the realm is still open. Anything else?

2. `File::try_remove` does indeed `throw PermissionDenied(msg, path);` (see https://github.com/realm/realm-core/blob/master/src/realm/util/file.cpp#L1381). My beginner C++ understanding is that this will be thrown all the way to the SDK, is that correct? So a throws comment should be added here. 👍 

3. I might be overengineering it now but I'd like to get the needs of all SDKs in here hence the following comparison:

| . | .NET | Cocoa | Java |
| ------------- | ------------- | ------------- | ------------- |
| Realm is open        | throws                                                  | returns false + error object | returns false |
| Files considered     | realm + .log_a / .log_b / .log / .lock / .note | realm + .note                       | realm + .note |
| File does not exist  | safe (ignored)                             | safe (ignored)                      | safe (ignored) |
| File not deleted      | throws if error                                   | returns false + error object  | logs a warning |
| Folders considered | .management                                      | .management                       | .management |
| Folder not found    |  safe (ignored)                            | safe (ignored)                      | safe (ignored) |
| Folder not deleted  | throws if error                                   | returns false + error object  | logs a warning |
| Return value           | void                                                      | true if realm file deleted, false otherwise + optional error object | true if realm file deleted, false if not deleted or realm is open |

To sum that up, what do you think about this combination of behaviours?

| . | Core | 
| ------------- | ------------- | 
| Realm is open        | throws `class RealmNotClosed: public std::runtime_error` |
| Files considered     | realm + .log_a / .log_b / .log / .note / .backup.realm |
| File does not exist  | safe (ignored) | 
| File not deleted      | throws if error | 
| Folders considered | .management |
| Folder not found    |  safe (ignored) |
| Folder not deleted  | throws if error | 
| Return value           | probably just void since it either deleted everything or throws |

Catching the thrown error and rethrow (.NET) / log warning (Java) / transform into an error object (Cocoa) would then almost exactly lead to the same behaviour in all SDKs except that they're using a unified OS / Core function that's also safe (even though to be fair, Java and Cocoa already call `DB::call_with_lock` themselved and don't touch the `.lock` file).

Reply:
b52c00bacfc82d059d135d0ffc17552d135d3ca7 should now correctly implement this. Along with 2d0fdd72f4951558ef1dee16d66e5cdf798455c9 for the comments.

Reply:
>  @return Yes, if all files could either be deleted or did not exist.     
No if at least one existing file coule not be deleted.

This comment still doesn't seem correct, because there's no way the function can return false. If this is intended, you may as well change the return value to void because it doesn't give you any extra information.

Reply:
This is also reworked, switched it to void now since I'm only returning true or an exception.